library/core/src/num/dec2flt/parse.rs

```diff
@@ -1,121 +1,246 @@
+    v -= 0x3030_3030_3030_3030;
+    v = (v * 10) + (v >> 8); // will not overflow, fits in 63 bits
+    let v1 = (v & MASK).wrapping_mul(MUL1);
+    let v2 = ((v >> 16) & MASK).wrapping_mul(MUL2);
+    ((v1.wrapping_add(v2) >> 32) as u32) as u64
 }
 
-#[derive(Debug, PartialEq, Eq)]
-/// The interesting parts of a decimal string.
-pub struct Decimal<'a> {
-    pub integral: &'a [u8],
-    pub fractional: &'a [u8],
-    /// The decimal exponent, guaranteed to have fewer than 18 decimal digits.
-    pub exp: i64,
+/// Parse digits until a non-digit character is found.
+fn try_parse_digits(s: &mut AsciiStr<'_>, x: &mut u64) {
+    // may cause overflows, to be handled later
+    s.parse_digits(|digit| {
+        *x = x.wrapping_mul(10).wrapping_add(digit as _);
+    });
 }
 
-impl<'a> Decimal<'a> {
-    pub fn new(integral: &'a [u8], fractional: &'a [u8], exp: i64) -> Decimal<'a> {
-        Decimal { integral, fractional, exp }
+/// Parse up to 19 digits (the max that can be stored in a 64-bit integer).
+fn try_parse_19digits(s: &mut AsciiStr<'_>, x: &mut u64) {
+    // SAFETY: safe since s cannot be empty when first and step are called.
+    unsafe {
+        while *x < MIN_19DIGIT_INT && !s.is_empty() && s.first_unchecked().is_ascii_digit() {
+            let digit = s.first_unchecked() - b'0';
+            *x = (*x * 10) + digit as u64; // no overflows here
+            s.step();
+        }
     }
 }
 
-#[derive(Debug, PartialEq, Eq)]
-pub enum ParseResult<'a> {
-    Valid(Decimal<'a>),
-    ShortcutToInf,
-    ShortcutToZero,
-    Invalid,
+/// Try to parse 8 digits at a time, using an optimized algorithm.
+fn try_parse_8digits(s: &mut AsciiStr<'_>, x: &mut u64) {
+    // may cause overflows, to be handled later
+    if let Some(v) = s.read_u64() {
+        if is_8digits(v) {
+            *x = x.wrapping_mul(1_0000_0000).wrapping_add(parse_8digits(v));
+            // SAFETY: already ensured the buffer was >= 8 bytes in read_u64.
+            unsafe {
+                s.step_by(8);
+            }
+            if let Some(v) = s.read_u64() {
+                if is_8digits(v) {
+                    *x = x.wrapping_mul(1_0000_0000).wrapping_add(parse_8digits(v));
+                    // SAFETY: already ensured the buffer was >= 8 bytes in try_read_u64.
+                    unsafe {
+                        s.step_by(8);
+                    }
+                }
+            }
+        }
+    }
 }
 
-/// Checks if the input string is a valid floating point number and if so, locate the integral
-/// part, the fractional part, and the exponent in it. Does not handle signs.
-pub fn parse_decimal(s: &str) -> ParseResult<'_> {
-    if s.is_empty() {
-        return Invalid;
+/// Parse the scientific notation component of a float.
+fn parse_scientific(s: &mut AsciiStr<'_>) -> i64 {
+    let start = *s;
+    // SAFETY: the first character is 'e'/'E' and scientific mode is enabled
+    unsafe {
+        s.step();
+    }
```

Comment:
This relies on a non-local safety invariant.
```suggestion
/// Parse the scientific notation component of a float.
unsafe fn parse_scientific(s: &mut AsciiStr<'_>) -> i64 {
    let start = *s;
    // SAFETY: the first character is 'e'/'E' and scientific mode is enabled
    unsafe {
        s.step();
    }
```

Reply:
[Fixed](https://github.com/Alexhuszagh/rust/blob/fb1a817fa13f774ce4da27fb9769db2779163030/library/core/src/num/dec2flt/parse.rs#L79-L103) by moving the exponent step logic out of `parse_scientific`. The logic now uses an `Option<i32>`, which then returns an invalid float if there is an exponent sign without any valid digits. If not, it short-circuits. Although this may not be strictly correct with a partial parser (we would want to simply set the cursor to the position before the exponent character), since core does not implement a partial parser, this short-circuit is fine.